crates/lib/src/execution_data.rs

```diff
@@ -65,44 +120,112 @@ impl ReadoutMap {
         readout_mappings: &HashMap<String, String>,
         readout_values: &HashMap<String, ReadoutValues>,
     ) -> Self {
-        let result = readout_values
-            .iter()
-            .flat_map(|(readout_name, values)| {
-                readout_mappings
-                    .iter()
-                    .filter(|(_, program_alias)| *program_alias == readout_name)
-                    .map(|(program_name, _)| program_name.as_ref())
-                    .map(parse_readout_register)
-                    .map(|reference| Ok((reference?, values.clone())))
-                    .collect::<Result<Vec<_>, MemoryReferenceParseError>>()
-            })
-            .flatten()
-            .collect::<HashMap<_, _>>()
-            .into();
+        let mut result = ReadoutMap(HashMap::new());
+        for (readout_name, values) in readout_values {
+            readout_mappings
+                .iter()
+                .filter_map(|(program_name, program_alias)| {
+                    (program_alias == readout_name).then(|| program_name.as_ref())
+                })
+                .map(parse_readout_register)
+                .filter_map(Result::ok)
+                .for_each(|reference| {
+                    let row = match &values.values {
+                        Some(readout_values::Values::IntegerValues(ints)) => ints
+                            .values
+                            .clone()
+                            .into_iter()
+                            .map(|i| Some(ReadoutValue::Integer(i)))
+                            .collect(),
+                        Some(readout_values::Values::ComplexValues(comps)) => comps
+                            .values
+                            .clone()
+                            .into_iter()
+                            .map(|c| Complex64::new(c.real().into(), c.imaginary().into()))
+                            .map(|c| Some(ReadoutValue::Complex(c)))
+                            .collect(),
+                        None => Vec::new(),
+                    };
+                    // TODO handle possible truncation
+                    let shape = (reference.index as usize + 1, row.len());
```

Comment:
@Shadow53 does it make sense for `quil-rs` to use `usize` for a memory index? Would keep us from having to bubble errors up here.

Reply:
I'm not the most familiar with the non-parsing bits of `quil-rs`, buuuuut....

I did a search for uses of `MemoryReference::index` in `quil-rs` and [this is the only time](https://github.com/rigetti/quil-rs/blob/e6e42bf04f34babd154f23ae81c599868fc6dbe2/src/expression.rs#L365) I see the field being used rather than set:

```rust
let value = values.get(memory_reference.index as usize)?;
```

Yeah, I'd say it makes sense. I would guess that the reason it is currently a `u64` is because the parser uses `Token::Integer(u64)`. Not sure if it would need a new `Token` variant for this use case, but that can be left for discussion in its own issue.

Reply:
cc @kalzoo and @notmgsk

Reply:
For now I recreated the `MemoryReference` struct that uses a `usize` for the index instead of `u64`. We were already doing the parsing of the reference as a string in the library, so it was simple enough to cast the index to a `usize` and let the existing error handling do the rest. 

`MemoryReference` parsing errors are ignored by the `from_mapping_and_values` method, so any `index` that exceeds `usize` just wouldn't be populated in the matrix. This seems like a nearly impossible edge case, how could one write and run a Quil program that exceeded their systems `usize`? Even on 32-bit systems, would anyone write a serious program with `2^32` indices on a register?

Reply:
See this comment for a reason why we might want to avoid `usize`: https://github.com/rigetti/quil-rs/issues/8#issuecomment-885265144

Reply:
That comment makes sense, but in practicality there is going to be *somewhere* that the value turns into a `usize`. In that case, one of two things must happen:

- Some arches have the value truncated (`as usize`)
- Some arches hit an error during conversion (`usize::try_from()`)

IMO, the most backwards-compatible solution without unexpected side effects (i.e. truncation) is to have a new `Token` type in `quil-rs` for indices specifically, and surface an error if the parsed index does not fit a `usize` for some reason.

This is something that should never happen in practice, but I'd prefer to surface an error if someone decides to try this anyways.

Reply:
So the [Quil spec](https://quil-lang.github.io/) doesn't specify the bitness of integers. However, along the lines of that comment that @notmgsk linked - I don't think the concerns of using these integers should leak backwards into the parser. 

Imagine that someone _does_ write a program with such a large register and happens to parse and run it on a 32-bit machine. Should that fail in parsing, or in usage with our client library? Of the two, the latter seems like the better choice. It's also the easiest to implement - rather than truncating here, I'd `try_from` ... although that will have to make this function fallible.

Reply:
Cool, that aligns with what I settled on for the PR. The memory reference parser now bubbles up errors and uses `usize::from_str()` which blows up if it would overflow the usize.